typescript-for-haskellers/article.md

```diff
@@ -0,0 +1,916 @@
+
+## Base concepts
+
+Unlike Haskell in TypeScript types are not the center part of the language.
+They just help programmers to write stricter and more understandable code, making their life easier.
+You can use types like `any` and nobody pushes you into strict frames.
+
+So, this part will be about simple built-in concepts which are familiar for haskellers and do not require any overheads on writing code.
+Nevertheless, we will also leave links for some more complex libraries and articles which may help you investigate the case even deeper.
+
+### Strictness
+
+By default, TypeScript is really polite in using types.
+It was developed in a such way to give an ability for JavaScript developers simply using it.
+Nevertheless, with bunch or compiler options you can make it more strict in types and other things.
+So, the main compiler option is which you need to add to your `tsconfig.json` file is `strict`.
+It will enable all strictness-related flags for you and if in future something will be added to strict options list you will not need to add it manually.
+But in case you want to specify them all, here you are: alwaysStrict, strictNullChecks, strictBindCallApply, strictFunctionTypes, strictPropertyInitialization, noImplicitAny, noImplicitThis, useUnknownInCatchVariables.
+
+### Type aliases and newtypes
+
+#### Type aliases
+
+Type aliases works exactly like they do in Haskell -- you can create a new name to refer to some type.
+
+```typescript
+type Email = string;
+```
+
+And also you are able to create types with properties.
+Symbol `?` here is used as a syntactic sugar for `string | undefined` type.
+```typescript
+type Document = {
+  name: string;
+  athor?: string;
+}
+```
+
+#### Newtype
+
+Creating a `newtype` is not so trivial.
+There is no such analog in TypeScript, but you can do it using tags.
+
+And here is how all this magic works.
+With intersection types (& syntax) you can add `tag` field for your type.
+Using `symbol` you are able to declare const-named properties on types.
+Word `unique` means that they are unique.
+Your type is tagged with unique symbol.
+
+```typescript
+type Email = string & { readonly __tag: unique symbol };
+type City = string & { readonly __tag: unique symbol };
+```
+
+Using `as` which means type assertion you are able to cast your base type to tagged type.
+And when you will try to pass value of another tagged type, compiler will say that two unique symbols are different.
+
+```typescript
+function sendMessage(email: Email);
+
+sendMessage("message"); // Argument of type 'string' is not assignable to parameter of type 'Email'.
+sendMessage("St. Petersburg" as City); // Argument of type 'City' is not assignable to parameter of type 'Email'.
+
+sendMessage("email@gmail.com" as Email); // Ok
+```
+
+Anyway, this is not an ideal solution for newtypes.
+There is a [comment](https://github.com/Microsoft/TypeScript/issues/4895#issuecomment-401067935) in issue with a better solution.
```

Comment:
If that one is better, I have a question as a reader as to why I was provided with a worse one?

Reply:
It is better for like complex system and it is also complex for this example.

Reply:
While this is the simplest solution for newtypes, it is not the best for complex systems. For a better (but more complex) one, check this [comment](https://github.com/Microsoft/TypeScript/issues/4895#issuecomment-401067935) in TypeScript's issues.

Reply:
I think this would sound better, then.